README.md
[![Steam Workshop Subscribers](https://img.shields.io/endpoint?style=for-the-badge&url=https%3A%2F%2Fshieldsio-steam-workshop.jross.me%2F2850494354%2Fsubscriptions-text)](https://steamcommunity.com/sharedfiles/filedetails/?id=2850494354)
[![Steam Workshop Favorites](https://img.shields.io/endpoint?style=for-the-badge&url=https%3A%2F%2Fshieldsio-steam-workshop.jross.me%2F2850494354%2Ffavourites-text)](https://steamcommunity.com/sharedfiles/filedetails/?id=2850494354)

# Bring the Boys Home

[![GitHub release](https://img.shields.io/github/release/msolefonte/tww3-btbh.svg?style=flat)](https://GitHub.com/msolefonte/tww3-btbh/releases/)
[![Patreon](https://img.shields.io/endpoint.svg?url=https%3A%2F%2Fshieldsio-patreon.vercel.app%2Fapi%3Fusername%3Dwolfylpdc%26type%3Dpatrons&style=flat)](https://www.patreon.com/wolfylpdc)
[![GitHub license](https://img.shields.io/github/license/msolefonte/tww3-btbh?style=flat)](https://github.com/msolefonte/tww3-btbh/blob/master/LICENSE)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat)](http://makeapullrequest.com)

[Steam Workshop Mod Link](https://steamcommunity.com/sharedfiles/filedetails/?id=2850494354)

Bring the Boys Home mod for Total War: Warhammer III.

## Introduction

Approved port of Jadawin's [Bring the Boys Home](https://steamcommunity.com/sharedfiles/filedetails/?id=2619922901) for
Warhammer II.

This mod introduces a new feature to the game: war exhaustion. This value, based on the duration of the war and on how
good is it going for the faction, is used to force AI to ask for peace once it has passed a threshold. This fixes a
classic problem with the game where the AI will never ask for help even if it is on a serious disadvantage.

Notable changes introduced from the original Jadawin version include AI to AI peace treaties, extended configuration via
Mod Configuration Tool and automatic support for support for DLC and mod races and factions. Performance and
optimization has also been improved.

## Contributing

Feel free to add issues or to create pull requests. Help is always welcome.

## Versioning

[SemVer](http://semver.org/) is used for versioning. For the changelog, see [CHANGELOG.md](CHANGELOG.md).

## Authors

* **Jadawin** - *Initial work* - [jadawin](https://steamcommunity.com/profiles/76561198030772148)
* **Marc Solé Fonte** - *Conversion to Warhammer III and posterior work* - [msolefonte](https://github.com/msolefonte)

## License

This project is licensed under the Apache License, Version 2.0 - see the [LICENSE](LICENSE) file for details.

## Acknowledgments

* Special thanks to **Jadawin** for his amazing work for Warhammer II and for being open about me porting the mod here.
Thank you for all your excellent mods!
* Special thanks to the **Da Modding Den** community for their knowledge and patience. This would have not not been
possible without them. I resist to give names because you are all amazing <3
* Special thanks to all the users that have supported the development economically.
* Thanks to all the users that have participated in the beta releases, that have reported bugs and that have contributed
with reports, feelings and new ideas. You keep this alive.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 5.10.2021

- Original work by Jadawin

## [2.0.0] - 18.08.2022

- Complete rework from scratch
- Added automatic support for DLC and mod races and factions
- Added support for AI to AI peace treaties
- Added player customization for different parameters
  - Thresholds for war duration, amount of battles, war exhaustion and minimum losses
  - Time between offers
- Added improved logging features for easier bug reports
- Modified war exhaustion formula to: `losses / (wins + losses) * 100 + duration - 5`


src/script/campaign/mod/a.lua
--logging
local function JADLOG(text)
  if not (__write_output_to_logfile or __enable_jadlog) then
    return;
  end
  local logText = tostring(text)
  local logTimeStamp = os.date("%d, %m %Y %X")
  local popLog = io.open("jadawin_bring_the_boys_home.txt","a")
  popLog :write("Bring the Boys Home: [".. logTimeStamp .. "]: "..logText .. " \n")
  popLog :flush()
  popLog :close()
end

local function JADSESSIONLOG()
  if not (__write_output_to_logfile or __enable_jadlog) then
    return;
  end
  local logTimeStamp = os.date("%d, %m %Y %X")
  local popLog = io.open("jadawin_bring_the_boys_home.txt","w+")
  popLog :write("NEW LOG ["..logTimeStamp.."] \n")
  popLog :flush()
  popLog :close()
end
JADSESSIONLOG()

local function jlog(text)
  JADLOG(tostring(text))
end

--general won a battle: count victories if it's either the player or an AI that defeated the player
core:add_listener(
  "JBTBH_LordWonBattle",
  "CharacterCompletedBattle",
  function(context)
    return (cm:character_is_army_commander(context:character()) and context:character():won_battle())
  end,
  function(context)
    local player_faction_key = cm:get_human_factions()[1]
    local player_faction = cm:model():world():faction_by_key(player_faction_key)
    local character = context:character()

    if character:faction():is_human() then
      jlog("JBTBH_LordWonBattle Human")
      local defeated_faction = "none"
      local defeated_side_strength = 0
      local turn_number = cm:model():turn_number()
      local loser_value_threshold = 3000 * math.ceil(turn_number / 25)
      if loser_value_threshold > 9000 then loser_value_threshold = 9000 end
      local last_battle_had_ai_garrison = false
      local last_battle = cm:model():pending_battle()
      --Was the battle at a settlement?
      if(last_battle:has_contested_garrison()) then
        local garrison_residence = last_battle:contested_garrison()
        if(not garrison_residence:faction():is_human()) then
          --jlog("Last battle included an AI garrison.")
          last_battle_had_ai_garrison = true
        end
      end

      if cm:pending_battle_cache_human_is_attacker() then
        --player was the attacker
        --jlog("Player was attacker.")
        if last_battle_had_ai_garrison and cm:pending_battle_cache_num_defenders() == 1 then
          jlog("Last battle was against only an AI garrison, does not count.")
          return
        end
        defeated_faction = jget_short(cm:pending_battle_cache_get_defender_faction_name(1))
        defeated_side_strength = cm:pending_battle_cache_defender_value()
      elseif  cm:pending_battle_cache_human_is_defender() then
        --player was the defender
        --jlog("Player was defender.")
        defeated_faction = jget_short(cm:pending_battle_cache_get_attacker_faction_name(1))
        defeated_side_strength = cm:pending_battle_cache_attacker_value()
      else
        jlog("Human was involved but neither attacker nor defender.")
      end
      if (defeated_faction ~= "none") and (defeated_side_strength >= loser_value_threshold) then
        jlog("Player lord won a battle against faction: "..(jget_long(defeated_faction)).." Value: "..(defeated_side_strength))
        if cm:get_saved_value("btbh_losses_"..(defeated_faction)) then
          local losses = cm:get_saved_value("btbh_losses_"..(defeated_faction))
          cm:set_saved_value("btbh_losses_"..(defeated_faction), (losses+1))
        else
          cm:set_saved_value("btbh_losses_"..(defeated_faction), 1)
        end
      elseif (defeated_faction ~= "none") and (defeated_side_strength < loser_value_threshold) then
        jlog("Defeated side was too weak for the battle to count. "..(defeated_side_strength))
      else
        jlog("Player lord won a battle but defeated faction is not valid")
      end
    elseif character:faction():at_war_with(player_faction) then
      --jlog("JBTBH_LordWonBattle AI at war with player")
      --battle winner is AI and at war with player
      if cm:pending_battle_cache_human_is_involved() then
        --AI that is at war with player was also actually fighting the player in the last battle
        jlog("AI lord won a battle against player.")
        local ai_faction_short = jget_short(context:character():faction():name())
        if (ai_faction_short ~= "none") then
          if cm:get_saved_value("btbh_wins_"..ai_faction_short) then
            local wins = cm:get_saved_value("btbh_wins_"..ai_faction_short)
            cm:set_saved_value("btbh_wins_"..ai_faction_short, (wins+1))
          else
            cm:set_saved_value("btbh_wins_"..ai_faction_short, 1)
          end
        end
      end
    end
  end,
  true
)

--Player TurnStart: Count up the war duration counter for all current wars
core:add_listener(
  "JBTBH_PlayerTurnStart",
  "FactionTurnStart",
  function(context) return (context:faction():is_human()) end,
  function(context)
    local turn_number = cm:model():turn_number()
    jlog("----------JBTBH_PlayerTurnStart ["..(turn_number).."] ----------")
    local player_faction = context:faction()
    local war_factions = player_faction:factions_at_war_with()
    local current_faction_key = ""
    local duration = false
    for i=0, war_factions:num_items()-1 do
      current_faction_key = war_factions:item_at(i):name()
      duration = false
      --jlog("Checking faction: "..(current_faction_key))
      if jget_short(current_faction_key) ~= "none" then
        duration = cm:get_saved_value("btbh_dur_"..(jget_short(current_faction_key)))
        if duration then
          duration = duration + 1
          cm:set_saved_value("btbh_dur_"..(jget_short(current_faction_key)), duration)
        else
          duration = 1
          cm:set_saved_value("btbh_dur_"..(jget_short(current_faction_key)), duration)
        end
        jlog("["..current_faction_key.."] War duration: "..duration)
      end
    end
  end,
  true
)

--PeaceTreatySigned: Reset war duration counters
core:add_listener(
  "JBTBH_PeaceTreatySigned",
  "PositiveDiplomaticEvent",
  function(context) return ((context:is_peace_treaty()) and (context:proposer():is_human() or context:recipient():is_human())) end,
  function(context)
    --jlog("JBTBH_PeaceTreatySigned")
    local ai_faction_key = "none"
    if not context:proposer():is_human() then
      ai_faction_key = context:proposer():name()
    elseif not context:recipient():is_human() then
      ai_faction_key = context:recipient():name()
    end
    if (jget_short(ai_faction_key) ~= "none") then
      --reset war duration
      jlog("Peace treaty signed. Reset war duration, wins and losses for "..ai_faction_key)
      cm:set_saved_value("btbh_dur_"..(jget_short(ai_faction_key)), false)
      cm:set_saved_value("btbh_wins_"..(jget_short(ai_faction_key)), false)
      cm:set_saved_value("btbh_losses_"..(jget_short(ai_faction_key)), false)
    else
      jlog("Cannot reset war duration because AI faction "..ai_faction_key.." is not valid.")
    end
  end,
  true
)

--WarDeclared: Reset war duration counters (in case a faction was destroyed but later comes back to life, so no peace treaty was ever signed)
core:add_listener(
  "JBTBH_WarDeclared",
  "NegativeDiplomaticEvent",
  function(context) return ((context:is_war()) and (context:proposer():is_human() or context:recipient():is_human())) end,
  function(context)
    --jlog("JBTBH_WarDeclared")
    local ai_faction_key = "none"
    if not context:proposer():is_human() then
      ai_faction_key = context:proposer():name()
    elseif not context:recipient():is_human() then
      ai_faction_key = context:recipient():name()
    end
    if (jget_short(ai_faction_key) ~= "none") then
      --reset war duration
      jlog("War declared. Reset war duration, wins and losses for "..ai_faction_key)
      cm:set_saved_value("btbh_dur_"..(jget_short(ai_faction_key)), false)
      cm:set_saved_value("btbh_wins_"..(jget_short(ai_faction_key)), false)
      cm:set_saved_value("btbh_losses_"..(jget_short(ai_faction_key)), false)
    else
      jlog("Cannot reset war duration because AI faction "..ai_faction_key.." is not valid.")
    end
  end,
  true
)


--AI TurnStart: Offer peace if conditions are met
core:add_listener(
  "JBTBH_AITurnStart",
  "FactionTurnStart",
  function(context) return not(context:faction():is_human()) end,
  function(context)
    local player_faction_key = cm:get_human_factions()[1]
    local player_faction = cm:model():world():faction_by_key(player_faction_key)
    local ai_faction = context:faction()
    local turn_number = cm:model():turn_number()
    if ai_faction:at_war_with(player_faction) then
      --jlog("JBTBH_AITurnStart At war with player")
      local war_duration = cm:get_saved_value("btbh_dur_"..(jget_short(ai_faction:name())))
      local wins_against_player = cm:get_saved_value("btbh_wins_"..(jget_short(ai_faction:name())))
      local losses_against_player = cm:get_saved_value("btbh_losses_"..(jget_short(ai_faction:name())))
      if not wins_against_player then wins_against_player = 0 end
      if not losses_against_player then losses_against_player = 0 end
      if war_duration then
        jlog((ai_faction:name()).." war data: Duration "..(war_duration).."| Wins "..(wins_against_player).."| Losses "..(losses_against_player).."|")
        if war_duration > 10 and ((wins_against_player + losses_against_player) >= 3) then
          local win_ratio = 1
          if losses_against_player > 0 then
            win_ratio = (wins_against_player / (wins_against_player + losses_against_player))
            jlog("Win ratio against player: "..(win_ratio).." in battles: "..(wins_against_player + losses_against_player))
            local turn_next_peace_offer = turn_number
            if cm:get_saved_value("btbh_wait_"..(jget_short(ai_faction:name()))) then
              turn_next_peace_offer = (cm:get_saved_value("btbh_wait_"..(jget_short(ai_faction:name()))) + 5)
            end
            if win_ratio < 0.3 and (turn_next_peace_offer <= turn_number) then
              cm:faction_offers_peace_to_other_faction(ai_faction:name(), player_faction_key)
              --set wait counter (current turn number) so AI does not sue for peace every turn
              cm:set_saved_value("btbh_wait_"..(jget_short(ai_faction:name())), turn_number)
            elseif win_ratio < 0.3 then
              jlog("AI would sue for peace but wait counter has not expired yet.")
            end
          end
        else
          --jlog("War duration ("..(war_duration)..") or number of wins/losses ("..(wins_against_player + losses_against_player)..") too low to consider suing for peace.")
        end
      else
        --jlog("No war duration stored for faction ["..(ai_faction:name()).."]. Either first turn of war or not an eligible faction.")
      end
    end
  end,
  true
)


src/script/campaign/mod/btbh.lua
local active_wars = {};
local config = {
  time_beetween_offers = 5,
  threshold_battles = 0,
  threshold_duration = 10,
  threshold_loss_value = 3000,
  threshold_war_exhaustion = 70,
  logging_enabled = false
};

-- GENERIC --

local function persist_table(table_name, t, callback)
    cm:add_saving_game_callback(function(context) cm:save_named_value("btbh_"..table_name, t, context) end);
    cm:add_loading_game_callback(function(context)
        local loaded_table = cm:load_named_value("btbh_"..table_name, t, context);
        callback(loaded_table);
    end);
end

local function get_config(config_key)
  if get_mct then
    local mct = get_mct();

    if mct ~= nil then
      local mod_cfg = mct:get_mod_by_key("wolfy_bring_the_boys_home");
      return mod_cfg:get_option_by_key(config_key):get_finalized_setting();
    end
  end

  return config[config_key];
end

local function btbh_log(str)
  if get_config("logging_enabled") then
    out("[wolfy][btbh] " .. str);
  end
end

-- ACTIVE WARS --

local function aw_create_entry(faction_key, enemy_faction_key)
  if active_wars[faction_key] == nil then
    active_wars[faction_key] = { [enemy_faction_key] = { duration = 0, wins = 0, losses = 0, last_offer = 0 } };
  else
    active_wars[faction_key][enemy_faction_key] = { duration = 0, wins = 0, losses = 0, last_offer = 0 };
  end
end

local function aw_delete_entry(faction_key, enemy_faction_key)
  if active_wars[faction_key] ~= nil then
    active_wars[faction_key][enemy_faction_key] = nil;

    local next = next;
    if next(active_wars[faction_key]) == nil then
      active_wars[faction_key] = nil;
    end
  end
end

local function aw_add_defeat(faction_key, enemy_faction_key)
  if active_wars[faction_key] == nil then
    active_wars[faction_key] = { [enemy_faction_key] = { duration = 0, wins = 0, losses = 1, last_offer = 0 } };
  elseif active_wars[faction_key][enemy_faction_key] == nil then
    active_wars[faction_key][enemy_faction_key] = { duration = 0, wins = 0, losses = 1, last_offer = 0 };
  else
    active_wars[faction_key][enemy_faction_key]["losses"] = active_wars[enemy_faction_key][faction_key]["losses"] + 1;
  end
end

local function aw_add_victory(faction_key, enemy_faction_key)
  if active_wars[faction_key] == nil then
    active_wars[faction_key] = { [enemy_faction_key] = { duration = 0, wins = 1, losses = 0, last_offer = 0 } };
  elseif active_wars[faction_key][enemy_faction_key] == nil then
    active_wars[faction_key][enemy_faction_key] = { duration = 0, wins = 1, losses = 0, last_offer = 0 };
  else
    active_wars[faction_key][enemy_faction_key]["wins"] = active_wars[faction_key][enemy_faction_key]["wins"] + 1;
  end
end

-- BRING THE BOYS HOME --

local function update_war_stats_after_battle(winner_faction)
  local winner_faction_key = winner_faction:name();
  local defeated_faction_key, defeated_side_strength;

  if cm:pending_battle_cache_faction_is_attacker(winner_faction_key) then
    defeated_faction_key = cm:pending_battle_cache_get_defender_faction_name(1);
    defeated_side_strength = cm:pending_battle_cache_defender_value();
  elseif cm:pending_battle_cache_faction_is_defender(winner_faction_key) then
    defeated_faction_key = cm:pending_battle_cache_get_attacker_faction_name(1);
    defeated_side_strength = cm:pending_battle_cache_attacker_value();
  else
    btbh_log("Faction " .. winner_faction_key .. " won a battle, but it is not attacker nor defender");
    return;
  end

  btbh_log("Faction " .. winner_faction_key .. " won a battle against " .. defeated_faction_key);
  if defeated_side_strength >= get_config("threshold_loss_value") then
    btbh_log("Defeated side strength is over threshold (" .. defeated_side_strength .. "). Proceeding.");
    aw_add_defeat(defeated_faction_key, winner_faction_key);
    aw_add_victory(winner_faction_key, defeated_faction_key);
  else
    btbh_log("Defeated side was too weak for the battle to count (" .. defeated_side_strength .. ")");
  end
end

local function update_all_active_wars_duration()
  btbh_log("Updating war duration for all active wars");
  for _, enemy_factions in pairs(active_wars) do
    for _, war_stats in pairs(enemy_factions) do
      war_stats["duration"] = war_stats["duration"] + 1;
    end
  end
end

local function clean_war_stats(faction_a, faction_b, reason)
  btbh_log("Reseting empty stats for " .. faction_a .. " vs " .. faction_b);
  aw_delete_entry(faction_a, faction_b);
  aw_delete_entry(faction_b, faction_a);
end

local function reset_war_stats(faction_a, faction_b, reason)
  btbh_log("Creating empty stats for " .. faction_a .. " vs " .. faction_b);
  aw_create_entry(faction_a, faction_b);
  aw_create_entry(faction_b, faction_a);
end

local function send_required_peace_treaty_requests(faction_key)
  if active_wars[faction_key] ~= nil then
    for enemy_faction_key, war_stats in pairs(active_wars[faction_key]) do
      local duration, wins, losses = war_stats["duration"], war_stats["wins"], war_stats["losses"];
      btbh_log(faction_key .. " vs " .. enemy_faction_key .. " war stats");
      btbh_log("Duration: " .. duration ..  " | Wins: " .. wins  .. " | Losses: " .. losses)
      if duration > get_config("threshold_duration") and wins + losses > get_config("threshold_battles") then
        local war_exhaustion =  losses / (wins + losses) * 100 + duration - 5;
        btbh_log("War Exhaustion: " .. war_exhaustion);

        if war_exhaustion > get_config("threshold_war_exhaustion") then
          local turn_number = cm:model():turn_number();
          if war_stats["last_offer"] <= turn_number - get_config("time_beetween_offers") then
            btbh_log("Faction " .. faction_key .. " will sue " .. enemy_faction_key .. " for peace");
            cm:faction_offers_peace_to_other_faction(faction_key, enemy_faction_key);
            war_stats["last_offer"] = turn_number;
          end
        end
      end
    end
  end
end

-- LISTENERS --

local function add_listeners()
  core:add_listener(
    "BTBH_LordWonBattle",
    "CharacterCompletedBattle",
    function(context)
      return cm:character_is_army_commander(context:character()) and context:character():won_battle();
    end,
    function(context)
      update_war_stats_after_battle(context:character():faction());
    end,
    true
  )

  core:add_listener(
    "BTBH_PeaceTreatySigned",
    "PositiveDiplomaticEvent",
    function(context)
      return context:is_peace_treaty();
    end,
    function(context)
      clean_war_stats(context:proposer():name(), context:recipient():name());
    end,
    true
  )

  core:add_listener(
    "BTBH_WarDeclared",
    "NegativeDiplomaticEvent",
    function(context)
      return context:is_war();
    end,
    function(context)
      reset_war_stats(context:proposer():name(), context:recipient():name());
    end,
    true
  )

  core:add_listener(
    "BTBH_PlayerTurnStart",
    "FactionTurnStart",
    function(context) return (context:faction():is_human()) end,
    update_all_active_wars_duration,
    true
  )

  core:add_listener(
    "BTBH_AITurnStart",
    "FactionTurnStart",
    function(context)
      return not context:faction():is_human();
    end,
    function(context)
      send_required_peace_treaty_requests(context:faction():name());
    end,
    true
  )
end

-- MAIN --

local function main()
  persist_table("active_wars", active_wars, function(t) active_wars = t end);
  add_listeners();
end

main();


src/script/mct/settings/bthb.lua
if not get_mct then return end
local mct = get_mct();

if not mct then return end
local mct_mod = mct:register_mod("wolfy_bring_the_boys_home");

mct_mod:set_title("Bring the Boys Home", false);
mct_mod:set_author("Wolfy & Jadawin");
mct_mod:set_description("No Endless Wars", false);

mct_mod:add_new_section("1-btbh-base", "Base Options", false);

local option_th_duration = mct_mod:add_new_option("threshold_duration", "slider");
option_th_duration:set_text("Minimum War Duration");
option_th_duration:set_tooltip_text("How many turns of war until start considering peace?");
option_th_duration:slider_set_min_max(1, 50);
option_th_duration:slider_set_step_size(1);
option_th_duration:set_default_value(10);

local option_th_battles = mct_mod:add_new_option("threshold_battles", "slider");
option_th_battles:set_text("Minimum War Battles");
option_th_battles:set_tooltip_text("How many battles completed until start considering peace?");
option_th_battles:slider_set_min_max(0, 20);
option_th_battles:slider_set_step_size(1);
option_th_battles:set_default_value(0);

local option_th_exhaustion = mct_mod:add_new_option("threshold_war_exhaustion", "slider");
option_th_exhaustion:set_text("Minimum War Exhaustion");
option_th_exhaustion:set_tooltip_text("How much war exhaustion should force a peace request?\n\n[[col:yellow]]" ..
                                      "War exhaustion formula: losses / (wins + losses) * 100 + duration - 5[[/col]]");
option_th_exhaustion:slider_set_min_max(0, 150);
option_th_exhaustion:slider_set_step_size(5);
option_th_exhaustion:set_default_value(70);

local option_th_value = mct_mod:add_new_option("threshold_loss_value", "slider");
option_th_value:set_text("Minimum Army Cost");
option_th_value:set_tooltip_text("How much should an army cost to be considered for a victory?");
option_th_value:slider_set_min_max(0, 20000);
option_th_value:slider_set_step_size(500);
option_th_value:set_default_value(3000);

local option_btbh_tbo = mct_mod:add_new_option("time_beetween_offers", "slider");
option_btbh_tbo:set_text("Time Between Offers");
option_btbh_tbo:set_tooltip_text("How many turns should AI wait before asking for peace again after a reject?");
option_btbh_tbo:slider_set_min_max(1, 100);
option_btbh_tbo:slider_set_step_size(1);
option_btbh_tbo:set_default_value(5);

mct_mod:add_new_section("2-btbh-ao", "Advanced Options", false);

local option_btbh_logging_enabled = mct_mod:add_new_option("logging_enabled", "checkbox");
option_btbh_logging_enabled:set_text("Enable logging");
option_btbh_logging_enabled:set_tooltip_text("If enabled, a log will be populated as you play. Use it to report bugs!");
option_btbh_logging_enabled:set_default_value(false);


src/script/campaign/mod/bkp.lua
local active_wars = {};
local config = {
  time_beetween_offers = 5;
  treshold_battles = 3,
  treshold_duration = 10,
  treshold_win_ratio = 30,
  logging_enabled = true
};

-- GENERIC --

local function persist_table(table_name, t, callback)
    cm:add_saving_game_callback(function(context) cm:save_named_value("btbh_"..table_name, t, context) end);
    cm:add_loading_game_callback(function(context)
        local loaded_table = cm:load_named_value("btbh_"..table_name, t, context);
        callback(loaded_table);
    end);
end

local function btbh_log(str)
  if get_config("logging_enabled") then
    out("[wolfy][btbh] " .. str);
  end
end

local function get_config(config_key)
  if get_mct then
    local mct = get_mct();

    if mct ~= nil then
      local mod_cfg = mct:get_mod_by_key("wolfy_bring_the_boys_home");
      return mod_cfg:get_option_by_key(config_key):get_finalized_setting();
    end
  end

  return config[config_key];
end

-- BRING THE BOYS HOME --

-- local function update_war_stats_after_battle(winner_character)
--   local player_faction_key = cm:get_human_factions()[1];  -- TODO What if MP
--   local player_faction = cm:model():world():faction_by_key(player_faction_key);
--
--   if winner_character:faction():is_human() then
--     local turn_number = cm:model():turn_number();
--     local loser_value_threshold = 3000 * math.ceil(turn_number / 25);
--     if loser_value_threshold > 9000 then loser_value_threshold = 9000 end
--
--     local defeated_faction = "none";
--     local defeated_side_strength = 0;
--
--     local last_battle_had_ai_garrison = false;
--     local last_battle = cm:model():pending_battle();
--     --Was the battle at a settlement?
--     if(last_battle:has_contested_garrison()) then
--       local garrison_residence = last_battle:contested_garrison()
--       if not garrison_residence:faction():is_human() then
--         --btbh_log("Last battle included an AI garrison.")
--         last_battle_had_ai_garrison = true
--       end
--     end
--
--     if cm:pending_battle_cache_human_is_attacker() then
--       --player was the attacker
--       --btbh_log("Player was attacker.")
--       if last_battle_had_ai_garrison and cm:pending_battle_cache_num_defenders() == 1 then
--         btbh_log("Last battle was against an AI garrison alone. It does not count.");
--         return;
--       end
--       defeated_faction = jget_short(cm:pending_battle_cache_get_defender_faction_name(1))
--       defeated_side_strength = cm:pending_battle_cache_defender_value()
--     elseif  cm:pending_battle_cache_human_is_defender() then
--       --player was the defender
--       --btbh_log("Player was defender.")
--       defeated_faction = jget_short(cm:pending_battle_cache_get_attacker_faction_name(1))
--       defeated_side_strength = cm:pending_battle_cache_attacker_value()
--     else
--       btbh_log("Human was involved but neither attacker nor defender.")
--     end
--
--     if (defeated_faction ~= "none") and (defeated_side_strength >= loser_value_threshold) then
--       btbh_log("Player lord won a battle against faction: "..defeated_faction.." Value: "..(defeated_side_strength))
--       if cm:get_saved_value("btbh_losses_"..(defeated_faction)) then
--         local losses = cm:get_saved_value("btbh_losses_"..(defeated_faction))
--         cm:set_saved_value("btbh_losses_"..(defeated_faction), (losses+1))
--       else
--         cm:set_saved_value("btbh_losses_"..(defeated_faction), 1)
--       end
--     elseif (defeated_faction ~= "none") and (defeated_side_strength < loser_value_threshold) then
--       btbh_log("Defeated side was too weak for the battle to count. "..(defeated_side_strength))
--     else
--       btbh_log("Player lord won a battle but defeated faction is not valid")
--     end
--   -- elseif winner_character:faction():at_war_with(player_faction) then -- TODO Remove so it works for MP
--   --   --btbh_log("BTBH_LordWonBattle AI at war with player")
--   --   --battle winner is AI and at war with player
--   --   if cm:pending_battle_cache_human_is_involved() then
--   --     --AI that is at war with player was also actually fighting the player in the last battle
--   --     btbh_log("AI lord won a battle against player")
--   --     local ai_faction_short = jget_short(chracter:faction():name())
--   --     if (ai_faction_short ~= "none") then
--   --       if cm:get_saved_value("btbh_wins_"..ai_faction_short) then
--   --         local wins = cm:get_saved_value("btbh_wins_"..ai_faction_short)
--   --         cm:set_saved_value("btbh_wins_"..ai_faction_short, (wins+1))
--   --       else
--   --         cm:set_saved_value("btbh_wins_"..ai_faction_short, 1)
--   --       end
--   --     end
--   --   end
--   else
--     btbh_log("AI lord won a battle")
--     local winner = get_shorter_name(winner_character:faction():name())
--     local loser = cm:pending_battle_cache_get_defender_faction_name(1)
--     if (ai_faction_short ~= "none") then
--       if cm:get_saved_value("btbh_wins_"..ai_faction_short) then
--         local wins = cm:get_saved_value("btbh_wins_"..ai_faction_short)
--         cm:set_saved_value("btbh_wins_"..ai_faction_short, (wins+1))
--       else
--         cm:set_saved_value("btbh_wins_"..ai_faction_short, 1)
--       end
--     end
--   end
-- end

local function update_all_active_wars_duration()
  btbh_log("Updating war duration for all active wars");
  for _, enemy_factions in pairs(active_wars) do
    for _, war_stats in pairs(enemy_factions) do
      war_stats["duration"] = war_stats["duration"] + 1;
    end
  end
end

local function _delete_active_war(faction_a, faction_b)
  if active_wars[faction_a] ~= nil then
    active_wars[faction_a][faction_b] = nil;

    local next = next;
    if next(active_wars[faction_a]) == nil then
      active_wars[faction_a] = nil;
    end
  end
end

local function clean_war_stats(faction_a, faction_b, reason)
  btbh_log("Reseting stats for " .. faction_a .. " vs " .. faction_b .. " (reason: " .. reason .. ")");
  _delete_active_war(faction_a, faction_b);
  _delete_active_war(faction_b, faction_a);
end

local function send_required_peace_treaty_requests(faction_name)
  if active_wars[faction_name] ~= nil then
    for enemy_faction_name, war_stats in pairs(active_wars[faction_name]) do
      local duration, wins, losses = war_stats["duration"], war_stats["wins"], war_stats["losses"];
      btbh_log(faction_name .. " vs " .. enemy_faction_name .. " war stats");
      btbh_log("Duration: " ..duration ..  " | Wins: " .. wins  .. " | Losses: " .. losses)
      if duration > get_config("treshold_duration") and wins + losses >= get_config("treshold_battles") then
        local win_ratio =  wins / (wins * losses) * 100;
        local turn_number = cm:model():turn_number();

        if win_ratio < get_config("treshold_win_ratio") and
           war_stats["last_offer"] <= turn_number - get_config("time_beetween_offers") then
          btbh_log("Faction " .. faction_name .. " will sue " .. enemy_faction_name .. " for peace");
          cm:faction_offers_peace_to_other_faction(faction_name, enemy_faction_name);
          war_stats["last_offer"] = turn_number;
        end
      end
    end
  end
end

-- LISTENERS --

local function add_listeners()
  --general won a battle: count victories if it's either the player or an AI that defeated the player
  core:add_listener(
    "BTBH_LordWonBattle",
    "CharacterCompletedBattle",
    function(context)
      return cm:character_is_army_commander(context:character()) and context:character():won_battle();
    end,
    function(context)
      update_war_stats_after_battle(context:character());
    end,
    true
  )

  core:add_listener(
    "BTBH_PeaceTreatySigned",
    "PositiveDiplomaticEvent",
    function(context)
      return context:is_peace_treaty();
    end,
    function(context)
      clean_war_stats(context:proposer():name(), context:recipient():name(), "Peace treaty signed");
    end,
    true
  )

  core:add_listener(
    "BTBH_WarDeclared",
    "NegativeDiplomaticEvent",
    function(context)
      return context:is_war();
    end,
    function(context)
      clean_war_stats(context:proposer():name(), context:recipient():name(), "War declared");
    end,
    true
  )

  core:add_listener(
    "BTBH_PlayerTurnStart",
    "FactionTurnStart",
    function(context) return (context:faction():is_human()) end,
    update_all_active_wars_duration,
    true
  )

  core:add_listener(
    "BTBH_AITurnStart",
    "FactionTurnStart",
    function(context)
      return not context:faction():is_human();
    end,
    function(context)
      send_required_peace_treaty_requests(context:faction():name());
    end,
    true
  )
end

-- MAIN --

local function main()
  persist_table("active_wars", active_wars, function(t) active_wars = t end);
  add_listeners();
end

main();
